stacks-signer/src/runloop.rs

```diff
@@ -13,1063 +13,211 @@
+        if needs_refresh {
+            if let Some(new_signer_config) = self.get_signer_config(reward_cycle) {
+                let signer_id = new_signer_config.signer_id;
+                debug!("Signer is registered for reward cycle {reward_cycle} as signer #{signer_id}. Initializing signer state.");
+                let prior_reward_cycle = reward_cycle.saturating_sub(1);
+                let prior_reward_set = prior_reward_cycle % 2;
+                if let Some(signer) = self.stacks_signers.get_mut(&prior_reward_set) {
+                    if signer.reward_cycle == prior_reward_cycle {
+                        // The signers have been calculated for the next reward cycle. Update the current one
+                        debug!("Signer #{}: Next reward cycle ({reward_cycle}) signer set calculated. Updating current reward cycle ({prior_reward_cycle}) signer.", signer.signer_id);
+                        signer.next_signer_ids = new_signer_config
+                            .registered_signers
+                            .signer_ids
+                            .values()
+                            .copied()
+                            .collect();
+                        signer.next_signer_slot_ids =
+                            new_signer_config.registered_signers.signer_slot_ids.clone();
+                    }
                 }
-            }
-            SignError::Aggregator(e) => {
-                warn!(
-                    "Signer #{}: Received an aggregator error: {:?}",
-                    self.signer_id, e
-                );
+                self.stacks_signers
+                    .insert(reward_index, Signer::from(new_signer_config));
+                debug!("Signer #{signer_id} for reward cycle {reward_cycle} initialized. Initialized {} signers", self.stacks_signers.len());
+            } else {
+                warn!("Signer is not registered for reward cycle {reward_cycle}. Waiting for confirmed registration...");
             }
         }
-        // TODO: should reattempt to sign the block here or should we just broadcast a rejection or do nothing and wait for the signers to propose a new block?
     }
 
-    /// Send any operation results across the provided channel
-    fn send_operation_results(
-        &mut self,
-        res: Sender<Vec<OperationResult>>,
-        operation_results: Vec<OperationResult>,
-    ) {
-        let nmb_results = operation_results.len();
-        match res.send(operation_results) {
-            Ok(_) => {
+    /// Refresh the signer configuration by retrieving the necessary information from the stacks node
+    /// Note: this will trigger DKG if required
+    fn refresh_signers(&mut self, current_reward_cycle: u64) -> Result<(), ClientError> {
+        let next_reward_cycle = current_reward_cycle.saturating_add(1);
+        self.refresh_signer_config(current_reward_cycle);
+        self.refresh_signer_config(next_reward_cycle);
+        // TODO: do not use an empty consensus hash
```

Comment:
Is there a ticket for this?

Reply:
There is now: https://github.com/stacks-network/stacks-core/issues/4420